### PR TITLE
Rework GroupMember and Contacts

### PR DESCRIPTION
**Pull request by SergeySeroshtan**
- Allow GroupMember to handle Contact.
- Allow CloudFileMember to handle Contact.

### CODE REVIEW

File: include/CloudFileMember.h [hunk 1]
```diff
@@ -47,27 +47,30 @@ class CloudFileMember
     enum class Type { Member, Owner };
 
     CloudFileMember();
-    CloudFileMember(const ContactHandler &contact, const Type type);
+    CloudFileMember(UserId memberId, Type type, ContactHandler contact = {});
```
Comment (fpohtmeh-github) [inline include/CloudFileMember.h]: Let's use 2 constructors:
`CloudFileMember(UserId memberId, Type type)`
`CloudFileMember(ContactHandler contact, Type type)`

File: include/CloudFileMember.h [hunk 2]
```diff
@@ -47,27 +47,30 @@ class CloudFileMember
...
-    Type type() const;
+    [[nodiscard]] UserId memberId() const;
+
+    [[nodiscard]] Type type() const;
+
+    [[nodiscard]] QString typeAsDisplayString() const;
+
+    [[nodiscard]] ContactHandler contact() const;
+
+    [[nodiscard]] CloudFileMember cloneWithContact(ContactHandler contact) const;
```
Comment (fpohtmeh-github) [inline include/CloudFileMember.h]: I think we can deal without cloning

File: include/messenger/GroupMember.h [hunk 3]
```diff
@@ -54,7 +54,7 @@ class GroupMember
     //
     GroupMember() = default;
 
-    GroupMember(GroupId groupId, UserId memberId, QString memberNickName, GroupAffiliation memberAffiliation);
+    GroupMember(GroupId groupId, UserId memberId, GroupAffiliation memberAffiliation, ContactHandler contact = {});
```
Comment (fpohtmeh-github) [inline include/messenger/GroupMember.h]: Let's use 2 constructors
`GroupMember(GroupId groupId, UserId memberId, GroupAffiliation memberAffiliation)`
`GroupMember(GroupId groupId, ContactHandler contact, GroupAffiliation memberAffiliation)`

File: src/models/ContactsModel.cpp [hunk 4]
```diff
@@ -169,10 +172,13 @@ void ContactsModel::loadAvatarUrl(const QString &contactUsername)
     }
 }
 
-void ContactsModel::setAvatarUrl(const ContactHandler contact, const QUrl &url)
+void ContactsModel::setAvatarUrl(const ContactHandler &contact, const QUrl &url)
 {
     if (const auto index = findByUsername(contact->username()); index.isValid()) {
-        m_contacts[index.row()]->setAvatarLocalPath(url.toLocalFile());
+        const auto &oldContact = m_contacts[index.row()];
+        auto newContact = std::make_unique<Contact>(*oldContact); // make copy
```
Comment (fpohtmeh-github) [inline src/models/ContactsModel.cpp:179]: Let's rework without copying

File: src/operations/ListMembersCloudFileOperation.cpp [hunk 5]
```diff
@@ -80,27 +88,34 @@ void Self::onListFetchErrorOccured(CloudFileRequestId requestId, const QString &
...
+        m_members.clear();
+
+        //
+        //  Emit update and finish.
+        //
         ListMembersCloudFileUpdate update;
-        update.parentFolder = m_parentFolder;
-        update.file = m_file;
-        update.members = m_members;
+        update.parentFolder = std::move(m_parentFolder);
```
Comment (fpohtmeh-github) [inline src/operations/ListMembersCloudFileOperation.cpp]: In general case, I don't think that reset of class members is a good move. I can't guarantee that they aren't called after reset.

**Review by fpohtmeh-github — COMMENTED** [review]
All members now refer to Contact. But processing of each member calls `findUserByUsername` and it slowdowns operation with members.
How about using of User inside member (GroupMember/CloudFileMember)?

File: include/CloudFileMember.h (hunk 2 above)
Comment (SergeySeroshtan) [reply include/CloudFileMember.h]: How?

File: include/CloudFileMember.h (hunk 1 above)
Comment (SergeySeroshtan) [reply include/CloudFileMember.h]: What the benefits?

File: include/messenger/GroupMember.h (hunk 3 above)
Comment (SergeySeroshtan) [reply include/messenger/GroupMember.h]: Agreed.

**Pull request merged**